### PR TITLE
Refactor mesos scheduler initialization

### PR DESCRIPTION
**Pull request by pschorf**
## Changes proposed in this PR
- Add a new method to compute cluster, `initialize`
- Move mesos scheduler and driver creation from `start-mesos-scheduler` to `initialize`
- Rename `start-mesos-scheduler` to `start-leader-selector`
- Remove `set-mesos-driver-hack!`

## Why are we making these changes?
This makes the initialization logic less mesos-specific, and also moves towards supporting multiple compute clusters (since you could call `initialize` on several clusters)

### CODE REVIEW

File: scheduler/src/cook/scheduler/scheduler.clj [hunk 1]
```diff
@@ -1552,17 +1442,9 @@
                         (assoc-in [:pool->offers-chan pool-name] offers-chan)
                         (assoc-in [:pool->resources-atom pool-name] resources-atom))))
                 {}
-                pools')
-        {:keys [batch-size]} progress-config
```
Comment (scrosby) [inline scheduler/src/cook/scheduler/scheduler.clj]: This removal/migration makes me a little nervous. How can we be sure that this all works?

File: scheduler/test/cook/test/scheduler/scheduler.clj [hunk 2]
```diff
@@ -1882,116 +1882,6 @@
             (is (= (:slave-id offer-1) (:slave-id task-1)))
             (is (= (:slave-id offer-2) (:slave-id task-2)))))))))
 
-(deftest test-in-order-status-update-processing
```
Comment (scrosby) [inline scheduler/test/cook/test/scheduler/scheduler.clj]: Why delete?

File: scheduler/src/cook/mesos/mesos_compute_cluster.clj [hunk 3]
```diff
@@ -14,14 +14,170 @@
...
+(meters/defmeter [cook-mesos scheduler mesos-error])
+(meters/defmeter [cook-mesos scheduler handle-framework-message-rate])
+(meters/defmeter [cook-mesos scheduler handle-status-update-rate])
+
+
+(defn create-mesos-scheduler
+  "Creates the mesos scheduler which processes status updates asynchronously but in order of receipt."
+  [gpu-enabled? conn heartbeat-ch pool->fenzo pool->offers-chan match-trigger-chan
+   handle-exit-code handle-progress-message sandbox-syncer-state compute-cluster]
+  (let [configured-framework-id (cook.config/framework-id-config)
```
Comment (scrosby) [inline scheduler/src/cook/mesos/mesos_compute_cluster.clj]: Something to either fix here or put in punch-list for later, but I believe that this cook.config/framework-id-config is one of the two remaining uses of this config variable. Might want to think about we're going to replace this.

One thought, we cast compute-cluster to MesosComputeCluster, then extract the :framework-id from it. THen this call goes away, and this code becomes (more) generic.

File: scheduler/test/cook/test/scheduler/scheduler.clj (hunk 2 above)
Comment (pschorf) [reply scheduler/test/cook/test/scheduler/scheduler.clj]: Moved it to another file, which I forgot to `git add` 😳

File: scheduler/src/cook/components.clj [hunk 4]
```diff
@@ -294,10 +280,27 @@
...
+                             trigger-chans]
+                         (let [constructor (util/lazy-load-var 'cook.mesos.mesos-compute-cluster/->MesosComputeCluster)
+                               create-mesos-compute-cluster (fn [compute-cluster-name framework-id db-id driver-atom]
+                                                              (constructor compute-cluster-name
+                                                                           framework-id
+                                                                           db-id
+                                                                           driver-atom
+                                                                           sandbox-syncer-state
+                                                                           exit-code-syncer-state
+                                                                           mesos-heartbeat-chan
```
[the remark this answers is not in the document]
Comment (scrosby) [reply scheduler/src/cook/components.clj]: Thats reasonable. Can you put a line item in the punchlist for this refactor work?

File: scheduler/src/cook/mesos.clj [hunk 5]
```diff
@@ -194,43 +157,39 @@
...
                                           :rebalancer-reservation-atom rebalancer-reservation-atom
-                                          :sandbox-syncer-state sandbox-syncer-state
                                           :task-constraints task-constraints
                                           :trigger-chans trigger-chans})
-                                        driver (make-mesos-driver-fn scheduler framework-id)]
-                                    (mesomatic.scheduler/start! driver)
-                                    (reset! current-driver driver)
-                                    (cc/set-mesos-driver-atom-hack! compute-cluster driver)
-
+                                        cluster-leadership-promise (cc/initialize-cluster compute-cluster
```
[the remark this answers is not in the document]
Comment (scrosby) [reply scheduler/src/cook/mesos.clj]: Something like 'We leave a thread blocked .takeLeadership. When that thread is unblocked by curator, it will subsequently send a message on this channel, unblocking us here."

File: scheduler/src/cook/mesos/mesos_compute_cluster.clj [hunk 6]
```diff
@@ -65,17 +259,18 @@
...
-      (atom nil))))
+  ([conn create-mesos-compute-cluster mesos-cluster]
+    (get-mesos-compute-cluster conn create-mesos-compute-cluster mesos-cluster nil))
+  ([conn create-mesos-compute-cluster {:keys [compute-cluster-name framework-id] :as mesos-cluster} driver]
+   {:pre [compute-cluster-name
+          framework-id]}
+   (let [cluster-entity-id (get-mesos-cluster-entity-id (d/db conn) mesos-cluster)]
+     (when-not cluster-entity-id
+       (cc/write-compute-cluster conn (mesos-cluster->compute-cluster-map-for-datomic mesos-cluster)))
+     (create-mesos-compute-cluster compute-cluster-name
```
[the remark this answers is not in the document]
Comment (scrosby) [reply scheduler/src/cook/mesos/mesos_compute_cluster.clj]: Still don't like. :(  But the least worst evil.

File: scheduler/src/cook/scheduler/scheduler.clj (hunk 1 above)
Comment (scrosby) [reply scheduler/src/cook/scheduler/scheduler.clj]: I'll let you be the judge of 'enough testing'. :)

File: scheduler/src/cook/components.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (pschorf) [reply scheduler/src/cook/components.clj]: Done

File: scheduler/src/cook/mesos/mesos_compute_cluster.clj (hunk 6 above)
[the remark this answers is not in the document]
Comment (pschorf) [reply scheduler/src/cook/mesos/mesos_compute_cluster.clj]: Yeah, hopefully some of this can be further cleaned up later. But I wanted to stick to moving things as much as possible in this PR and avoid changing too much.

File: scheduler/src/cook/mesos.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (pschorf) [reply scheduler/src/cook/mesos.clj]: Did you see the comment below?

File: scheduler/src/cook/compute_cluster.clj [hunk 7]
```diff
@@ -16,8 +16,8 @@
 (ns cook.compute-cluster
   (:require [clojure.tools.logging :as log]
             [cook.config :as config]
-            [datomic.api :as d]
-            [mesomatic.scheduler :as mesos]))
+            [cook.datomic]
```
Comment (shamsimam) [inline scheduler/src/cook/compute_cluster.clj]: Why is this required?

File: scheduler/src/cook/components.clj [hunk 8]
```diff
@@ -294,10 +280,27 @@
...
+                                                                           trigger-chans))]
+                           ((util/lazy-load-var 'cook.mesos.mesos-compute-cluster/setup-compute-cluster-map-from-config) datomic/conn settings
+                             create-mesos-compute-cluster)))
      :mesos-datomic-mult (fnk []
                            (first ((util/lazy-load-var 'cook.datomic/create-tx-report-mult) datomic/conn)))
+     :mesos-heartbeat-chan (fnk []
```
Comment (shamsimam) [inline scheduler/src/cook/components.clj]: Can we add a todo that this mesos-specific stuff will be scoped down?

File: scheduler/src/cook/compute_cluster.clj [hunk 9]
```diff
@@ -28,10 +28,14 @@
...
+    "Initializes the cluster. Returns a channel that will be delivered on when the cluster loses leadership.")
+
+  (current-leader? [this]
+    "Returns true if this cook instance is currently the leader for the compute cluster")
+
   (get-mesos-driver-hack [this]
-    "Get the mesos driver. Hack; any funciton invoking this should be put within the compute-cluster implementation")
-  (set-mesos-driver-atom-hack! [this driver]
-    "Hack to overwrite the driver. Used until we fix the initialization order of compute-cluster"))
+    "Get the mesos driver. Hack; any funciton invoking this should be put within the compute-cluster implementation"))
```
Comment (shamsimam) [inline scheduler/src/cook/compute_cluster.clj]: typo in `funciton`

Comment (shamsimam) [reply scheduler/src/cook/compute_cluster.clj]: We should add a TODO here to clearly flag it as something that needs refactoring.

File: scheduler/src/cook/compute_cluster.clj [hunk 10]
```diff
@@ -28,10 +28,14 @@
   (db-id [this]
     "Get a database entity-id for this compute cluster (used for putting it into a task structure).")
 
+  (initialize-cluster [this pool->fenzo pool->offers-chan]
+    "Initializes the cluster. Returns a channel that will be delivered on when the cluster loses leadership.")
```
Comment (shamsimam) [inline scheduler/src/cook/compute_cluster.clj]: Should we worry about when the cluster regains leadership? Or do we expect the cluster to exit when it loses leadership? Either way, should we clarify in the docstring?

Comment (shamsimam) [reply scheduler/src/cook/compute_cluster.clj]: We should clarify what the channel values can be, e.g. an exception is supported as a return value and `nil` when there was no error.

File: scheduler/src/cook/mesos/mesos_compute_cluster.clj [hunk 11]
```diff
@@ -14,14 +14,169 @@
...
+   mesos-master             -- str, (optional) connection string for mesos masters
+   scheduler                -- mesos scheduler implementation
+   framework-id             -- str, (optional) Id of framework if it has connected to mesos before
+   "
+  ([config scheduler]
+   (make-mesos-driver config scheduler nil))
+  ([{:keys [mesos-framework-name mesos-role mesos-principal gpu-enabled?
+            mesos-failover-timeout mesos-master] :as config}
+    scheduler framework-id]
+   (apply mesomatic.scheduler/scheduler-driver
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/mesos_compute_cluster.clj]: Do we need to use `apply` here?

File: scheduler/src/cook/test/testutil.clj [hunk 12]
```diff
@@ -38,17 +38,31 @@
...
+  [compute-cluster-name framework-id db-id driver-atom]
+  (mcc/->MesosComputeCluster compute-cluster-name
+                             framework-id
+                             db-id
+                             driver-atom
+                             nil ; sandbox-syncer-state
+                             nil ; exit-code-syncer-state
+                             nil ; mesos-heartbeat-chan
+                             nil ; trigger-chans
+                             ))
```
Comment (shamsimam) [inline scheduler/src/cook/test/testutil.clj]: I find this easier to read:
```
(let [sandbox-syncer-state nil
      exit-code-syncer-state nil
      mesos-heartbeat-chan nil
      trigger-chans nil]
  (mcc/->MesosComputeCluster compute-cluster-name
                             framework-id
                             db-id
                             driver-atom
                             sandbox-syncer-state
                             exit-code-syncer-state
                             mesos-heartbeat-chan
                             trigger-chans))
```

File: scheduler/src/cook/scheduler/scheduler.clj [hunk 13]
```diff
@@ -1552,17 +1442,9 @@
...
     (start-jobs-prioritizer! conn pool-name->pending-jobs-atom task-constraints rank-trigger-chan)
-    {:scheduler (create-mesos-scheduler gpu-enabled? conn heartbeat-ch pool-name->fenzo pool->offers-chan
-                                        match-trigger-chan handle-exit-code handle-progress-message sandbox-syncer-state
-                                        compute-cluster)
-     :view-incubating-offers (fn get-resources-atom [p]
-                               (deref (get pool->resources-atom p)))}))
+    {:view-incubating-offers (fn get-resources-atom [p]
+                               (deref (get pool->resources-atom p)))
+     :pool->offers-chan pool->offers-chan
+     :pool-name->fenzo pool-name->fenzo}))
```
Comment (shamsimam) [inline scheduler/src/cook/scheduler/scheduler.clj]: Nitpick: lexicographic order

File: scheduler/src/cook/components.clj (hunk 8 above)
Comment (pschorf) [reply scheduler/src/cook/components.clj]: Done

File: scheduler/src/cook/compute_cluster.clj (hunk 7 above)
Comment (pschorf) [reply scheduler/src/cook/compute_cluster.clj]: Leftover from an old commit, removed.

File: scheduler/src/cook/compute_cluster.clj (hunk 10 above)
Comment (pschorf) [reply scheduler/src/cook/compute_cluster.clj]: We don't expect to regain leadership, since we'll keep the existing pattern of giving up cook leadership if we lose compute cluster leadership.

Added both to the docstring.

File: scheduler/src/cook/compute_cluster.clj (hunk 9 above)
Comment (pschorf) [reply scheduler/src/cook/compute_cluster.clj]: Fixed typo, added todo.

File: scheduler/src/cook/mesos/mesos_compute_cluster.clj (hunk 11 above)
Comment (pschorf) [reply scheduler/src/cook/mesos/mesos_compute_cluster.clj]: Done

File: scheduler/src/cook/scheduler/scheduler.clj (hunk 13 above)
Comment (pschorf) [reply scheduler/src/cook/scheduler/scheduler.clj]: Done

File: scheduler/src/cook/test/testutil.clj (hunk 12 above)
Comment (pschorf) [reply scheduler/src/cook/test/testutil.clj]: Done

**Comment by pschorf** [comment]
@shamsimam there was some nuance to that apply call, I made another commit to fix it.
